### PR TITLE
Add non-blocking JsonDiskCache and use it for msalias validation

### PR DESCRIPTION
**Pull request by yufeih**
https://dev.azure.com/ceapex/Engineering/_workitems/edit/144985/

The current msgraph and github cache are sync blocking cache, when items expires, the caller is blocked until the underlying API returns, this could starve threadpool.

This PR creates a `JsonDiskCache` that internally is a non-blocking async cache, and uses this cache against msgraph API. Due to the complexity of github user cache, it'll be handled in next PR.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5321)

### CODE REVIEW

File: src/docfx/lib/cache/JsonDiskCache.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                if (value.Expiry != null && value.Expiry < DateTime.UtcNow)
+                {
+                    // When the item expired, trigger background update but don't wait for the result
+                    _backgroundUpdates.GetOrAdd(key, UpdateDelegate(key, valueFactory)).Value.GetAwaiter();
+                }
+                return (default, value);
+            }
+
+            return _initialRequests.GetOrAdd(key, UpdateDelegate(key, valueFactory)).Value.GetAwaiter().GetResult();
```
Comment (live1206) [inline src/docfx/lib/cache/JsonDiskCache.cs]: seems `_initialRequests` is not needed? just `return UpdateDelegate(key, valueFactory).Value.GetAwaiter().GetResult();`

File: src/docfx/lib/cache/JsonDiskCache.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                    foreach (var cacheKey in item.GetKeys())
+                    {
+                        _cache.TryAdd(cacheKey, item);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets or update the cache asynchroniously.
```
Comment (live1206) [inline src/docfx/lib/cache/JsonDiskCache.cs]:
```suggestion
        /// Get or update the cache asynchronously.
```

File: src/docfx/lib/cache/JsonDiskCache.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            if (_needUpdate)
+            {
+                var content = JsonUtility.Serialize(new { items = _cache.Values });
+
+                Directory.CreateDirectory(Path.GetDirectoryName(_cachePath));
+                ProcessUtility.WriteFile(_cachePath, content);
+                _needUpdate = false;
+            }
+
+            return result.Select(item => item.error).ToArray();
```
Comment (live1206) [inline src/docfx/lib/cache/JsonDiskCache.cs]: filter out the null errors from result

File: src/docfx/lib/cache/JsonDiskCache.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/lib/cache/JsonDiskCache.cs]: This is to ensure only one call to `valueFactory` is made when multiple concurrent calls to `GetOrAdd` burst in for the same key.

But `_initialRequests` and `_backgroundUpdates` can be merged into one

File: src/docfx/lib/cache/JsonDiskCache.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                    foreach (var cacheKey in item.GetKeys())
+                    {
+                        _cache.TryAdd(cacheKey, item);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets an item from the cache asynchroniously, or creates the value if it does not exist.
```
Comment (live1206) [inline src/docfx/lib/cache/JsonDiskCache.cs]: typo: asynchroniously -> asynchronously

File: src/docfx/lib/cache/JsonDiskCache.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        /// The <paramref name="valueFactory"/> can be a long running asynchronious call,
```
Comment (live1206) [inline src/docfx/lib/cache/JsonDiskCache.cs]: typo

**Pull request merged**